zkevm-circuits/src/copy_circuit.rs

```diff
@@ -225,23 +228,35 @@ impl<F: Field> SubCircuitConfig<F> for CopyCircuitConfig<F> {
                     rw_diff,
                 );
             });
-            cb.condition(
-                and::expr([
-                    meta.query_advice(is_last, Rotation::cur()),
-                    tag.value_equals(CopyDataType::RlcAcc, Rotation::cur())(meta),
-                ]),
-                |cb| {
-                    cb.require_equal(
-                        "value == rlc_acc at the last row for RlcAcc",
-                        meta.query_advice(value, Rotation::cur()),
-                        meta.query_advice(rlc_acc, Rotation::cur()),
-                    );
-                },
-            );
 
             cb.gate(meta.query_fixed(q_enable, Rotation::cur()))
         });
 
+        meta.create_gate(
+            "Last Step (check value accumulator) Memory => Bytecode or RlcAcc",
+            |meta| {
+                let mut cb = BaseConstraintBuilder::default();
+
+                cb.require_equal(
+                    "value_acc_rlc == rlc_acc on the last row",
+                    meta.query_advice(value_acc_rlc, Rotation::next()),
+                    meta.query_advice(rlc_acc, Rotation::next()),
+                );
+
+                cb.gate(and::expr([
+                    meta.query_fixed(q_enable, Rotation::cur()),
+                    meta.query_advice(is_last, Rotation::next()),
+                    or::expr([
+                        and::expr([
+                            tag.value_equals(CopyDataType::Memory, Rotation::cur())(meta),
+                            tag.value_equals(CopyDataType::Bytecode, Rotation::next())(meta),
+                        ]),
+                        tag.value_equals(CopyDataType::RlcAcc, Rotation::next())(meta),
+                    ]),
```

Comment:
The expression from here may be causing the final gate degree ti increase more than 9 (and that's why tests are not passing).  A simple solution would be to split this into 2 constraints. Following this pattern:
`if A or B then X` -> `if A then X`, `if B then X`.

Another thing that comes to my mind is that `or::expr` returns a boolean, but to build a selector expression just having `0` when false and `!= 0` when true is enough, so perhaps you could replace the `or` by a `+`.  This would give `2` when both expressions are true, but that's ok.

Reply:
Nice idea! It looks nice and clean now, fixed in [52d480a](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1419/commits/52d480a1dcade244606b6d2598abd1f756a80c1d)